### 1. BASIC PROBABILITY


#### Câu hỏi 1: Kết quả của đoạn chương trình tính mean sau đây là:
Cho Data X = {2, 0, 2, 2, 7, 4, −2, 5, −1, −1}
Hãy hoàn thiện function compute_mean() để tính mean μ của X đã cho.

- Data: X = {x1, ..., xN }

- Mean: μ = $\frac{1}{N}*\displaystyle\sum_{i=1}^N x_i$

In [2]:
### Question 1
import numpy as np


def compute_mean(X):
	# your code here *********************

	return np.mean(X)


X = [2, 0, 2, 2, 7, 4, -2, 5, -1, -1]

print(" Mean : ", compute_mean(X))

 Mean :  1.8


Đáp án 1 là a) 1.8

#### Câu hỏi 2: Kết quả của đoạn chương trình tính median sau đây là: 
Cho data X= {1,5,4,4,9,13}. Hãy hoàn thiện function compute_median() để tìm median của X đã cho. 

- Data: X= {x1.....xn}
##### Median:
 - Sort X -> S (Tăng dần)
 - if N is odd: $m = S_{\frac{N+1}{2}}$
 - if N is even: $m = \frac{1}{2}*(S_{\frac{N}{2}}+S_{\frac{N}{2}+1})$

In [10]:
### Question 2

def compute_median(X):
	size = len(X)
	X = np.sort(X)
	print(X)
	if size % 2 == 0:
		return (X[size // 2 - 1] + X[size // 2]) / 2
	else:
		return X[size // 2]


X = [1, 5, 4, 4, 9, 13]
print(" Median : ", compute_median(X))

[ 1  4  4  5  9 13]
6
 Median :  4.5


Vậy kết quả là b) 4.5

Câu hỏi 3: Kết quả của đoạn chương tình tính variance và standard deviation sau đây là: 
Cho Data X={171,176,155,167,169,182}
Hãy hoàn thiện function compute_std() để tìm standard deviation σ của X đã cho

- Data: $X={x1,....,x_N}$
- Mean: $\mu = \frac{1}{N}*\displaystyle\sum_{i=1}^N x_i$
- Variance: $var(X) = \frac{1}{N}*\displaystyle\sum_{i=1}^N (x_i - \mu)^2 $
- Standard Deviation: $\sigma = \sqrt{var(X)}$

In [12]:
### Question 3

def compute_std(X):
	mean = compute_mean(X)
	variance = 0
	# your code here *******************
	N = len(X)
	variance = np.sum([(element - mean) ** 2 for element in X]) / N
	return np.around(np.sqrt(variance), 2)


X = [171, 176, 155, 167, 169, 182]
print(compute_std(X))

8.33


=> Đáp án là c) 8.33

#### Câu hỏi 4: Kết quả của đoạn chương trình tính correlation coeficient sau đây là:
Cho Data X = {-2,-5,-11,6,4,15,9}

Y = {4,25,121,36,16,225,81}
Hoàn thiện function compute_correlation_cofficient() để tìm correlation coefficient của X và Y đã cho :

- Random variables $X, Y: {x1,...,x_N} Y={y1,....,y_N}$
- Correlation Coefficient: $p_{xy} = \frac{E|(X-\mu_x)(Y-\mu_y)}{\sigma_x\sigma_y} = \frac{N(\sum{x_i y_i}) - (\sum{x_i})(\sum{y_i})}{\sqrt{N \sum{x_i^2} - (\sum{x_i})^2} \sqrt{N \sum{y_i^2} - (\sum{y_i})^2}}$

In [13]:
def compute_correlation_cofficient(X, Y):
	N_X = len(X)
	numerator = 0
	denominator = 0
	# your code here ****************
	mean_X = compute_mean(X)
	mean_Y = compute_mean(Y)
	std_X = compute_std(X)
	std_Y = compute_std(Y)
	numerator = sum((xi - mean_X) * (yi - mean_Y) for xi, yi in zip(X, Y)) / N_X
	denominator = std_X * std_Y
	return np.round(numerator / denominator, 2)


X = np.asarray([-2, -5, -11, 6, 4, 15, 9])
Y = np.asarray([4, 25, 121, 36, 16, 225, 81])
print(" Correlation : ", compute_correlation_cofficient(X, Y))

 Correlation :  0.42


Đáp án là D) 0.42